system/system_monitor/src/voltage_monitor/voltage_monitor.cpp

```diff
@@ -0,0 +1,168 @@
+  if (voltage_string_ == "") {
+    sensors_exists_ = false;
+  } else {
+    // Check if command exists
+    fs::path p = bp::search_path("sensors");
+    sensors_exists_ = (p.empty()) ? false : true;
+  }
+  gethostname(hostname_, sizeof(hostname_));
+  auto callback = &VoltageMonitor::checkBatteryStatus;
+  if (sensors_exists_) {
+    callback = &VoltageMonitor::checkVoltage;
+  }
+  updater_.add("CMOS Battery Status", this, callback);
+}
+
+static float getVoltage(std::string voltage_string)
+{
+  bp::ipstream is_out;
+  bp::ipstream is_err;
+  fs::path p = bp::search_path("sensors");
+  bp::child c(p.string(), bp::std_out > is_out, bp::std_err > is_err);
+  c.wait();
+
+  if (RCUTILS_UNLIKELY(c.exit_code() != 0)) {  // failed to execute sensors
+    return 0;
+  }
+  std::string line;
+  std::regex re(R"((\d+).(\d+))");  //    3.06 V  (min =  +0.00 V, max =  +4.08 V)
+  for (int i = 0; i < 200 && std::getline(is_out, line); i++) {
+    auto voltageStringPos = line.find(voltage_string.c_str());
+    if (voltageStringPos != std::string::npos) {
+      std::smatch match;
+      std::regex_search(line, match, re);
+      auto voltageString = match.str();
+      return std::stof(voltageString);
+    }
+  }
+  return 0;  // failed to read voltage
+}
+
+void VoltageMonitor::checkVoltage(diagnostic_updater::DiagnosticStatusWrapper & stat)
+{
+  // Remember start time to measure elapsed time
+  const auto t_start = SystemMonitorUtility::startMeasurement();
+
+  if (RCUTILS_UNLIKELY(!sensors_exists_)) {
+    stat.summary(DiagStatus::ERROR, "sensors error");
+    stat.add(
+      "sensors",
+      "Command 'sensors' not found, but can be installed with: 'sudo apt install lm-sensors' and "
+      "'sudo sensors-detect'");
+    return;
+  }
+
+  auto v = getVoltage(voltage_string_);
+
+  stat.add("CMOS battey voltage", fmt::format("{}", v));
+  if (RCUTILS_UNLIKELY(v < voltage_warn_)) {
+    stat.summary(DiagStatus::WARN, "LOW BATTERY");
+  } else {
+    stat.summary(DiagStatus::OK, "OK");
+  }
+
+  // Measure elapsed time since start time and report
+  SystemMonitorUtility::stopMeasurement(t_start, stat);
+}
+
+void VoltageMonitor::checkBatteryStatus(diagnostic_updater::DiagnosticStatusWrapper & stat)
+{
+  // Remember start time to measure elapsed time
+  const auto t_start = SystemMonitorUtility::startMeasurement();
+
+  // Get status of RTC
+  bp::ipstream is_out;
+  bp::ipstream is_err;
+  bp::child c("cat /proc/driver/rtc", bp::std_out > is_out, bp::std_err > is_err);
+  c.wait();
+
+  if (c.exit_code() != 0) {
+    std::ostringstream os;
+    is_err >> os.rdbuf();
+    stat.summary(DiagStatus::ERROR, "rtc error");
+    stat.add("rtc", os.str().c_str());
+    return;
+  }
+
+  std::string line;
+  bool status = false;
+  for (int i = 0; i < 200 && std::getline(is_out, line); i++) {
+    auto batStatusLine = line.find("batt_status");
+    if (batStatusLine != std::string::npos) {
+      auto batStatus = line.find("okay");
```

Comment:
Do you know what `batt_status` shows other than `okay` for example, if an error occurs?

Reply:
The code for rtc as follow.So the return value is okay or dead.
```:
	return seq_printf(seq,
			"periodic_IRQ\t: %s\n"
			"update_IRQ\t: %s\n"
			"HPET_emulated\t: %s\n"
			// "square_wave\t: %s\n"
			"BCD\t\t: %s\n"
			"DST_enable\t: %s\n"
			"periodic_freq\t: %d\n"
			"batt_status\t: %s\n",
			(rtc_control & RTC_PIE) ? "yes" : "no",
			(rtc_control & RTC_UIE) ? "yes" : "no",
			is_hpet_enabled() ? "yes" : "no",
			// (rtc_control & RTC_SQWE) ? "yes" : "no",
			(rtc_control & RTC_DM_BINARY) ? "no" : "yes",
			(rtc_control & RTC_DST_EN) ? "yes" : "no",
			cmos->rtc->irq_freq,
			(valid & RTC_VRT) ? "okay" : "dead"); <- "okay" or "dead"
```

Reply:
Thanks! Understood.